Use rustup.rs instead of custom bootstrap

SimonSapin opened the pull request:
Fixes #11361, closes #18874, fixes #19365.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/19395)
<!-- Reviewable:end -->

SimonSapin committed:
Use rustup.rs instead of custom bootstrap

Fixes #11361, closes #18874

SimonSapin committed:
Add rustup.rs instructions in README

SimonSapin committed:
Print a link to instructions if rustup is not found.

highfive commented:
Heads up! This PR modifies the following files:
 * @wafflespeanut: python/servo/devenv_commands.py, python/servo/post_build_commands.py, python/servo/command_base.py, python/servo/testing_commands.py, python/servo/util.py and 2 more
 * @rillian: rust-stable-version

SimonSapin commented:
Before this can land:

* A new version of rustup that includes https://github.com/rust-lang-nursery/rustup.rs/pull/1295 needs to be released (https://github.com/rust-lang-nursery/rustup.rs/issues/1275#issuecomment-346615370)
* That version needs to be install on CI builders

At first, contributors will probably have an older version of rustup and see something like this:

```
$ ./mach build
error: Found argument '--install' which wasn't expected, or isn't valid in this context

USAGE:
    rustup run <toolchain> <command>...

For more information try --help
Build FAILED in 0:00:00
```

This is pretty bad since `--help` in the output refers to `rustup run --help`, not `./mach build --help`. And even `rustup run --help` wouldn’t be helpful since it doesn’t know about options added in later versions.

Mach cannot easily distinguish this case from a "normal" build failure, so the only way to detect this situation and print a better diagnostic we would need to run `rustup --version` and parse its output every  time we’re about to run `rustup run`. This seems overkill, but perhaps I shouldn’t worry about that.

r? @jdm

SimonSapin committed:
Print a better error message when rustup is too old

SimonSapin commented:
`subprocess.check_output(["rustup", "--version"])` takes ~3ms on my laptop, so really not worth worrying about compared to existing overhead of Mach. I’ve added the detection mentioned above.

SimonSapin commented:
With this PR running `cargo build --manifest-path ports/servo/Cargo.toml` will use the same compiler as `./mach build`. However going back and forth between those will still recompile everything because mach sets the `RUSTFLAGS` environment variable.

To not have `RUSTFLAGS` by default we’d need rustc to revert back to using ld.gold by default: https://github.com/rust-lang/rust/pull/29974, and us to replace `-W unused-extern-crates` with `#![warn(unused_extern_crate)]` in each crate.

jdm commented:
Our Appvey and TravisCI are missing rustup; test-tidy is failing; and test-wpt is broken because ensure_bootstrapped was removed.

SimonSapin commented:
Thanks. I fixed these immediate issues, but as I mentioned we’re still waiting on a new rustup release.

jdm commented on python/servo/bootstrap_commands.py:
```diff
@@ -282,51 +125,31 @@ def bootstrap_pub_suffix(self, force=False):
+                        ['git', 'show', '%s:%s' % (commit_hash, toolchain_file)])
+                    old_toolchains.append(toolchain.strip())
+
+        remove_anything = False
```
This doesn't match the `remove_anything` variable later.

SimonSapin replied:
This as reported by tidy and fixed.

jdm commented on appveyor.yml:
```diff
@@ -40,6 +40,13 @@ cache:
+  - rustup-init.exe -y --default-host x86_64-pc-windows-msvc --default-toolchain none
+  - set PATH=%PATH%;C:\Users\appveyor\.cargo\bin
+  - rustc -V
+  - cargo -V
```
Should we modify the cache instructions above this block?

SimonSapin replied:
It looks like this caches the `.servo` directory including the extracted rustc and cargo downloaded by mach. Do we want to cache toolchains downloaded by rustup?

jdm replied:
Yes, since it makes builds a bit faster.

SimonSapin replied:
Alright, done.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #19410) made this pull request unmergeable. Please resolve the merge conflicts.

SimonSapin committed:
Cache rustup toolchains on AppVeyor and Travis-CI

SimonSapin committed:
Fix "rustup is not installed" message

bors-servo commented:
:umbrella: The latest upstream changes (presumably #19532) made this pull request unmergeable. Please resolve the merge conflicts.

SimonSapin commented:
Rustup release is blocked on https://github.com/rust-lang-nursery/rustup.rs/issues/1305

SimonSapin committed:
Use rustup "proxies" instead of `rustup run`

To make sure we’re indeed running rustup’s proxy
rather than some other `cargo` for example,
run the `rustup` executable with a different `argv[0]`.

SimonSapin commented:
Added a commit to unblock by not using the new rustup feature.

@bors-servo try

bors-servo commented:
:hourglass: Trying commit e44fdf76d0fab0e7ccb078d3129aeb3f80e04a04 with merge 91a9ad6817a28199d731ff306adc8b82155680bf...

bors-servo commented:
:broken_heart: Test failed - [arm32](http://build.servo.org/builders/arm32/builds/10340)

SimonSapin commented:
```
./mach clean-nightlies --keep 3 --force
[…]
 environment:
  PATH=/home/servo/bin:/usr/local/sbin:/usr/local/bin:/usr/bin:/usr/sbin:/sbin:/bin
[…]
OSError: [Errno 2] No such file or directory

  File "/home/servo/buildbot/slave/arm32/build/python/servo/bootstrap_commands.py", line 144, in clean_nightlies
    stdout = subprocess.check_output(['rustup', 'toolchain', 'list'])
```

`/home/servo/.cargo/bin` is not in `$PATH`. I think rustup normally adds it by writing to `~/.profile`, but maybe that file is not used in the environment where buildbot jobs run?

SimonSapin commented:
This effectively runs `subprocess.call(["cargo.exe", "build"], executable="rustup.exe")` which fails on AppVeyor with `WindowsError(2, 'The system cannot find the file specified')`.

Based on reading the Python standard library, this ends up calling `CreateProcess("rustup.exe", "cargo.exe build", …)`. Per https://msdn.microsoft.com/en-us/library/windows/desktop/ms682425(v=vs.85).aspx the `PATH` environment variable is *not* use for the first parameter, only for the second.

So the only way to run `rustup.exe` on Windows with a different `argv[0]` is to do the `PATH` lookup ourselves. I don’t know if there’s a better way to do that than parsing `PATH` and looking which files exist ourselves.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #19539) made this pull request unmergeable. Please resolve the merge conflicts.

SimonSapin commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit d8bab3e6c4f77b6011a9d6437a1b60139327a7c6 with merge 8242fda56c85a11e2ef14d562eae3c9b6140e4cb...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/7247)

SimonSapin commented:
@aneeshusa, does https://github.com/servo/saltfs/pull/762 still need another step to be deployed to CI builders?

jdm commented:
Yes, we need to highstate the builders.

SimonSapin committed:
Revert "Use rustup "proxies" instead of `rustup run`"

This reverts commit ad22368646290510149d33a0a41fc436ac4a0e96.

SimonSapin commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 30d962bcf76eb3a3a186553af2e4e38d034ef071 with merge 24a867ffb0afbf026fb2ea4d0f27c90be6713c93...

bors-servo commented:
:broken_heart: Test failed - [android](http://build.servo.org/builders/android/builds/10173)

SimonSapin commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 7f883d82ce07fd6762d9a82cce7cb67a747596e9 with merge a484ce4286e610356ea4ed82043a9410ac6a18a8...

bors-servo commented:
:broken_heart: Test failed - [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/5833)

SimonSapin committed:
rustup.rs: Use MSVC rather than GNU toolchains on Windows

SimonSapin commented:
On AppVeyor:

```rust
error: linking with `link.exe` failed: exit code: 1120
  |
  = note: "C:\\Program Files (x86)\\Microsoft Visual Studio 14.0\\VC\\BIN\\amd64\\link.exe" "/NOLOGO" "/NXCOMPAT" "/LARGEADDRESSAWARE" "/SAFESEH" "/LIBPATH:C:\\Users\\appveyor\\.rustup\\toolchains\\nightly-2017-11-14-i686-pc-windows-msvc\\lib\\rustlib\\i686-pc-windows-msvc\\lib" "C:\\projects\\servo\\target\\debug\\deps\\dom_struct-224c198e23df7fda.dom_struct0.rcgu.o" "C:\\projects\\servo\\target\\debug\\deps\\dom_struct-224c198e23df7fda.dom_struct1.rcgu.o" "C:\\projects\\servo\\target\\debug\\deps\\dom_struct-224c198e23df7fda.dom_struct2.rcgu.o" "C:\\projects\\servo\\target\\debug\\deps\\dom_struct-224c198e23df7fda.dom_struct3.rcgu.o" "/OUT:C:\\projects\\servo\\target\\debug\\deps\\dom_struct-224c198e23df7fda.dll" "/DEF:C:\\Users\\appveyor\\AppData\\Local\\Temp\\1\\rustc.Yl6vda9bXmAa\\lib.def" "C:\\projects\\servo\\target\\debug\\deps\\dom_struct-224c198e23df7fda.crate.metadata.rcgu.o" "C:\\projects\\servo\\target\\debug\\deps\\dom_struct-224c198e23df7fda.crate.allocator.rcgu.o" "/OPT:REF,NOICF" "/DEBUG" "/NATVIS:C:\\Users\\appveyor\\.rustup\\toolchains\\nightly-2017-11-14-i686-pc-windows-msvc\\lib\\rustlib\\etc\\liballoc.natvis" "/NATVIS:C:\\Users\\appveyor\\.rustup\\toolchains\\nightly-2017-11-14-i686-pc-windows-msvc\\lib\\rustlib\\etc\\libcore.natvis" "/LIBPATH:C:\\projects\\servo\\target\\debug\\deps" "/LIBPATH:C:\\Users\\appveyor\\.rustup\\toolchains\\nightly-2017-11-14-i686-pc-windows-msvc\\lib\\rustlib\\i686-pc-windows-msvc\\lib" "/LIBPATH:C:\\Users\\appveyor\\.rustup\\toolchains\\nightly-2017-11-14-i686-pc-windows-msvc\\lib\\rustlib\\i686-pc-windows-msvc\\lib" "proc_macro-11829f83756c64ad.dll.lib" "/LIBPATH:C:\\Users\\appveyor\\.rustup\\toolchains\\nightly-2017-11-14-i686-pc-windows-msvc\\lib\\rustlib\\i686-pc-windows-msvc\\lib" "syntax-0f21f5e615c5859c.dll.lib" "/LIBPATH:C:\\Users\\appveyor\\.rustup\\toolchains\\nightly-2017-11-14-i686-pc-windows-msvc\\lib\\rustlib\\i686-pc-windows-msvc\\lib" "rustc_errors-1b22ef7d4be6ca6a.dll.lib" "/LIBPATH:C:\\Users\\appveyor\\.rustup\\toolchains\\nightly-2017-11-14-i686-pc-windows-msvc\\lib\\rustlib\\i686-pc-windows-msvc\\lib" "syntax_pos-7d059d588bbad9ae.dll.lib" "/LIBPATH:C:\\Users\\appveyor\\.rustup\\toolchains\\nightly-2017-11-14-i686-pc-windows-msvc\\lib\\rustlib\\i686-pc-windows-msvc\\lib" "rustc_cratesio_shim-5a0ed2b459381b6f.dll.lib" "/LIBPATH:C:\\Users\\appveyor\\.rustup\\toolchains\\nightly-2017-11-14-i686-pc-windows-msvc\\lib\\rustlib\\i686-pc-windows-msvc\\lib" "term-3095e146c74f9c19.dll.lib" "/LIBPATH:C:\\Users\\appveyor\\.rustup\\toolchains\\nightly-2017-11-14-i686-pc-windows-msvc\\lib\\rustlib\\i686-pc-windows-msvc\\lib" "rustc_data_structures-12c636f652fe56e8.dll.lib" "/LIBPATH:C:\\Users\\appveyor\\.rustup\\toolchains\\nightly-2017-11-14-i686-pc-windows-msvc\\lib\\rustlib\\i686-pc-windows-msvc\\lib" "serialize-29271e07662eeebe.dll.lib" "/LIBPATH:C:\\Users\\appveyor\\.rustup\\toolchains\\nightly-2017-11-14-i686-pc-windows-msvc\\lib\\rustlib\\i686-pc-windows-msvc\\lib" "std-14232f553fbb2f9e.dll.lib" "C:\\Users\\appveyor\\.rustup\\toolchains\\nightly-2017-11-14-i686-pc-windows-msvc\\lib\\rustlib\\i686-pc-windows-msvc\\lib\\libcompiler_builtins-c3dbf71fb0229f85.rlib" "kernel32.lib" "advapi32.lib" "ws2_32.lib" "userenv.lib" "shell32.lib" "msvcrt.lib" "/DLL" "/IMPLIB:C:\\projects\\servo\\target\\debug\\deps\\dom_struct-224c198e23df7fda.dll.lib"
  = note:    Creating library C:\projects\servo\target\debug\deps\dom_struct-224c198e23df7fda.dll.lib and object C:\projects\servo\target\debug\deps\dom_struct-224c198e23df7fda.dll.exp
          dom_struct-224c198e23df7fda.dom_struct0.rcgu.o : error LNK2019: unresolved external symbol ___CxxFrameHandler3 referenced in function ___ehhandler$_ZN4core3ptr13drop_in_place17h042bfa3a294a8463E
          dom_struct-224c198e23df7fda.dom_struct1.rcgu.o : error LNK2001: unresolved external symbol ___CxxFrameHandler3
          dom_struct-224c198e23df7fda.dom_struct2.rcgu.o : error LNK2001: unresolved external symbol ___CxxFrameHandler3
          dom_struct-224c198e23df7fda.dom_struct3.rcgu.o : error LNK2001: unresolved external symbol ___CxxFrameHandler3
          dom_struct-224c198e23df7fda.dom_struct1.rcgu.o : error LNK2019: unresolved external symbol _memcpy referenced in function __ZN33_$LT$alloc..vec..Vec$LT$T$GT$$GT$3pop17h1e5fccfad93324e0E
          dom_struct-224c198e23df7fda.dom_struct2.rcgu.o : error LNK2001: unresolved external symbol _memcpy
          dom_struct-224c198e23df7fda.dom_struct3.rcgu.o : error LNK2001: unresolved external symbol _memcpy
          dom_struct-224c198e23df7fda.dom_struct3.rcgu.o : error LNK2019: unresolved external symbol _memset referenced in function __ZN5alloc6string6String4push17h89a721a5832f9b9bE
          LINK : error LNK2001: unresolved external symbol __DllMainCRTStartup@12
          LINK : error LNK2001: unresolved external symbol __load_config_used
          C:\Program Files (x86)\Windows Kits\10\lib\10.0.14393.0\um\x64\kernel32.lib : warning LNK4272: library machine type 'x64' conflicts with target machine type 'X86'
          C:\Program Files (x86)\Windows Kits\10\lib\10.0.14393.0\um\x64\advapi32.lib : warning LNK4272: library machine type 'x64' conflicts with target machine type 'X86'
          C:\Program Files (x86)\Windows Kits\10\lib\10.0.14393.0\um\x64\ws2_32.lib : warning LNK4272: library machine type 'x64' conflicts with target machine type 'X86'
          C:\Program Files (x86)\Windows Kits\10\lib\10.0.14393.0\um\x64\userenv.lib : warning LNK4272: library machine type 'x64' conflicts with target machine type 'X86'
          C:\Program Files (x86)\Windows Kits\10\lib\10.0.14393.0\um\x64\shell32.lib : warning LNK4272: library machine type 'x64' conflicts with target machine type 'X86'
          C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\LIB\amd64\msvcrt.lib : warning LNK4272: library machine type 'x64' conflicts with target machine type 'X86'
          C:\projects\servo\target\debug\deps\dom_struct-224c198e23df7fda.dll : fatal error LNK1120: 5 unresolved externals
          
error: aborting due to previous error
error: Could not compile `dom_struct`.

Caused by:
  process didn't exit successfully: `rustc --crate-name dom_struct components\dom_struct\lib.rs --crate-type proc-macro --emit=dep-info,link -C prefer-dynamic -C codegen-units=4 -C debuginfo=2 -C metadata=224c198e23df7fda -C extra-filename=-224c198e23df7fda --out-dir C:\projects\servo\target\debug\deps -L dependency=C:\projects\servo\target\debug\deps -W unused-extern-crates` (exit code: 101)

SimonSapin commented:
Remaining buildbot failures are:

* On OS X `~/.cargo/bin` is not in `$PATH` (https://github.com/servo/saltfs/pull/762). Additionally, as far as can tell rustup in installed at that location on some builders but not others (https://github.com/servo/saltfs/pull/758/files and https://github.com/servo/saltfs/pull/767).
* On Windows `C:\Users\Administrator\.cargo\bin` is in `$PATH` but rustup is apparently not installed.

SimonSapin commented:
On AppVeyor:

```
mach build -d -v && mach test-unit
info: syncing channel updates for 'nightly-2017-11-14-i686-pc-windows-msvc'
```
and then:
```
 = note: "C:\\Program Files (x86)\\Microsoft Visual Studio 14.0\\VC\\BIN\\amd64\\link.exe"  […]
```

So it looks like we’re using a 64-bit linker with a 32-bit compiler.

SimonSapin commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 01b4674cdc916c6b7e34a4857a0ffa68bf9dc4c6 with merge 71a37a631041995f4e5ec61d930aa13da1cdbe3e...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/2981)

SimonSapin commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 950a75a7ddf60b048021eccc4f575ab2e0dab66c with merge b771b95e2fca6b92c7b2572a7da27254bbe696ad...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/3074)

SimonSapin commented:
@bors-servo try retry

bors-servo commented:
:hourglass: Trying commit 950a75a7ddf60b048021eccc4f575ab2e0dab66c with merge 0315b38f92fc38dc99df6e6b843368ed7de9f0c1...

SimonSapin commented:
Ok, so now `$PATH` looks correct:

```
PATH=/home/servo/.cargo/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin
```

But calling rustup from mach still doesn’t work:

```python

OSError: [Errno 2] No such file or directory

  File "/Users/servo/buildbot/slave/mac-rel-wpt4/build/python/servo/bootstrap_commands.py", line 144, in clean_nightlies
    stdout = subprocess.check_output(['rustup', 'toolchain', 'list'])
```

It does work when I run the same Python code via SSH, though:

```sh
% ssh root@servo-mac2.servo.org
servo-mac2:~ root# su - servo
servo-mac2:~ servo$ echo $PATH
/Users/servo/.cargo/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin
servo-mac2:~ servo$ which rustup
/Users/servo/.cargo/bin/rustup
servo-mac2:~ servo$ rustup -V
rustup 1.8.0 (cb2d14862 2017-12-17)
servo-mac2:~ servo$ python
>>> import subprocess
>>> subprocess.check_output(['rustup', 'toolchain', 'list'])
'no installed toolchains\n'
```

SimonSapin commented:
Ah, it should be `/Users` in `PATH` instead of `/home`. Looks like `{{ common.servo_home }}` in https://github.com/servo/saltfs/pull/771/files is wrong…

SimonSapin commented:
`common/map.jinja`

```python
{%
  set common = salt['grains.filter_by']({
      'defaults': {
        'servo_home': '/home/servo',
      },
      'MacOS': {
        'servo_home': '/Users/servo'
      }
    },
    base='defaults',
    grain='os'
  )
%}
```

@aneeshusa, does this look right?

SimonSapin commented:
@bors-servo try

SimonSapin commented:
@bors-servo try retry

bors-servo commented:
:hourglass: Trying commit 950a75a7ddf60b048021eccc4f575ab2e0dab66c with merge 865990b70e1a72e1b7801ad5c4c65bd5b4b86cf6...

bors-servo commented:
:broken_heart: Test failed - [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/5980)

SimonSapin commented:
https://github.com/servo/saltfs/pull/758#discussion_r160287667

SimonSapin commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit e91a393d01650a2896a1da0f6fe77e6019aabc1a with merge 2099889c272c03716f70233a96be5e38fd061a59...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/3116)

SimonSapin commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit b2f90cfaed1837dd6dbc8ee89284df1685b2f1fc with merge 1ae8b6dd76977ee9148327161f4135c75392f4b0...

bors-servo commented:
:broken_heart: Test failed - [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/5985)

SimonSapin commented:
All non-Windows jobs are now green \o/

SimonSapin commented:
@KiChjang This is https://github.com/servo/saltfs/issues/773.

SimonSapin commented:
@bors-servo try retry

bors-servo commented:
:hourglass: Trying commit b2f90cfaed1837dd6dbc8ee89284df1685b2f1fc with merge f5135c2d99c833846e1a9dba23a9e64e137581e6...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/7500)

jdm commented on python/servo/post_build_commands.py:
```diff
@@ -211,10 +209,13 @@ def rr_replay(self):
+        rustc_path = check_output(["rustup" + BIN_SUFFIX, "which", "rustc"], env=env)
+        assert path.basename(path.dirname(rustc_path)) == "bin"
+        toolchain_path = path.dirname(path.dirname(rustc_path))
+        rust_docs = path.join(toolchain_path, "share", "doc", "rust", "html")
```
My only question is whether `./mach cargo-gecko doc` behaves as expected with these changes, since we use it in upload_docs.sh.

jdm reviewed (approved):
My only question is whether `./mach cargo-gecko doc` behaves as expected with these changes, since we use it in upload_docs.sh.

jdm commented:
That test failure is #19033.

SimonSapin commented:
Hmm, I get "rustdoc: Unrecognized option: 'document-private-items'" errors with `mach cargo-geckolib doc` but I wonder why they didn’t occur on master. Anyway, reverting d96f0ff6a77e309682898cc7e1905a7bd0feb3cf (until we upgrade the stable toolchain to a recent enough version) fixes it.

jdm commented:
Oh yeah, I forgot that https://github.com/servo/servo/issues/19365 has existed for multiple months now :(

jdm commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit be1a48c has been approved by `jdm`

<!-- @bors-servo r=jdm be1a48c4c468979b7c2a0bbbd1885f7e32f96c2f -->

bors-servo commented:
:hourglass: Testing commit be1a48c4c468979b7c2a0bbbd1885f7e32f96c2f with merge 4c4baa42481f26a2427c8800ff77ebc6aec291c6...

bors-servo commented:
:broken_heart: Test failed - [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/10704)

SimonSapin committed:
Revert "Use rustdoc --document-private-item instead of deprecated flags"

This reverts commit d96f0ff6a77e309682898cc7e1905a7bd0feb3cf.

SimonSapin commented:
```
KeyError: u'RUSTDOCFLAGS'

  File "/Users/servo/buildbot/slave/mac-dev-unit/build/python/servo/testing_commands.py", line 259, in test_unit
    del env["RUSTDOCFLAGS"]
```

Replaced `del` with `pop()`.

@bors-servo r=jdm

bors-servo commented:
:pushpin: Commit 2c33671 has been approved by `jdm`

<!-- @bors-servo r=jdm 2c3367182ecc888e45abd8a800d6a0ff800c77f8 -->

bors-servo commented:
:hourglass: Testing commit 2c3367182ecc888e45abd8a800d6a0ff800c77f8 with merge a618adea6840d56efdf1b6be951106f057e92d53...

bors-servo commented:
:broken_heart: Test failed - [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/6001)

jdm commented:
```
   Doc-tests script_tests
error: An unknown error occurred

To learn more, run the command again with --verbose.
```

jdm commented on python/servo/command_base.py:
```diff
@@ -582,41 +571,6 @@ def handle_android_target(self, target):
-
-        # Always check if all needed MSVC dependencies are installed
-        if "msvc" in target_platform:
-            Registrar.dispatch("bootstrap", context=self.context)
```
Given the appveyor failure (`thread 'main' panicked at 'OpenSSL library directory does not exist: C:\projects\servo\.servo\msvc-dependencies\openssl\1.1.0e-vs2015\lib64', C:\Users\appveyor\.cargo\registry\src\github.com-1ecc6299db9ec823\openssl-sys-0.9.22\build.rs:69:8`), I fear we may have lost something important by removing this line.

SimonSapin replied:
Reverted.

SimonSapin commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 2206e9725840caa6ce790efb1b2e5b5bd5520068 with merge e1f2fad931d79906345fa34a03c04ca771f8290a...

SimonSapin commented:
Ugh, the `RUSTDOCFLAGS` revert causes warnings in `./mach test-unit` for every crate that has doctests enabled:

```
WARNING: the 'no-defaults' flag is considered deprecated
WARNING: please see https://github.com/rust-lang/rust/issues/44136
WARNING: the 'passes' flag is considered deprecated
WARNING: please see https://github.com/rust-lang/rust/issues/44136
WARNING: (you may want to use --document-private-items)
```

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/10633)

SimonSapin commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 41c3dd2bad2d937c62c23f0754e5ee38d92567cd with merge d033780042fed7fbfe40a7f3644dcb13654198f9...

bors-servo commented:
:broken_heart: Test failed - [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/6009)

SimonSapin commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 272fef3117c33ba759c7d3812ccb78d5bf3e60d6 with merge f26eb55b9938cf3499e8b0005b042686dd47f964...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/10349), [arm32](http://build.servo.org/builders/arm32/builds/10585), [arm64](http://build.servo.org/builders/arm64/builds/10611), [linux-dev](http://build.servo.org/builders/linux-dev/builds/10637), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/7498), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/7513), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/10717), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/3122), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/3142), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/7391), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/7453), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/3212), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/3162), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/6014)
State: approved= try=True

SimonSapin commented:
@bors-servo r=jdm

bors-servo commented:
:pushpin: Commit 0681e68 has been approved by `jdm`

<!-- @bors-servo r=jdm 0681e682e999ed7ec7c7067ac1b788290544ceab -->

bors-servo commented:
:hourglass: Testing commit 0681e682e999ed7ec7c7067ac1b788290544ceab with merge 2be49ac80659f58f2a39910dd79a3f3b5da98945...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/10352), [arm32](http://build.servo.org/builders/arm32/builds/10588), [arm64](http://build.servo.org/builders/arm64/builds/10614), [linux-dev](http://build.servo.org/builders/linux-dev/builds/10640), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/7501), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/7516), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/10720), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/3125), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/3145), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/7394), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/7456), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/3215), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/3165), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/6017)
Approved by: jdm
Pushing 2be49ac80659f58f2a39910dd79a3f3b5da98945 to master...
